lib/AST/DiagnosticEngine.cpp

```diff
@@ -189,22 +201,31 @@ InFlightDiagnostic &InFlightDiagnostic::fixItRemove(SourceRange R) {
     charRange = CharSourceRange(charRange.getStart(),
                                 charRange.getByteLength()+1);
   }
-  Engine->getActiveDiagnostic().addFixIt(Diagnostic::FixIt(charRange, {}));
+  Engine->getActiveDiagnostic().addFixIt(Diagnostic::FixIt(charRange, {}, {}));
   return *this;
 }
 
+InFlightDiagnostic &
+InFlightDiagnostic::fixItReplace(SourceRange R, StringRef FormatString,
+                                 ArrayRef<DiagnosticArgument> Args) {
+  auto &SM = Engine->SourceMgr;
+  auto charRange = toCharSourceRange(SM, R);
 
-InFlightDiagnostic &InFlightDiagnostic::fixItReplace(SourceRange R,
-                                                     StringRef Str) {
-  if (Str.empty())
-    return fixItRemove(R);
+  Engine->getActiveDiagnostic().addFixIt(
+      Diagnostic::FixIt(charRange, FormatString, std::move(Args)));
+  return *this;
+}
 
-  assert(IsActive && "Cannot modify an inactive diagnostic");
-  if (R.isInvalid() || !Engine) return *this;
+InFlightDiagnostic &InFlightDiagnostic::fixItReplace(SourceRange R, StringRef Str) {
+  if (Str.empty())
+     return fixItRemove(R);
 
+   assert(IsActive && "Cannot modify an inactive diagnostic");
+   if (R.isInvalid() || !Engine) return *this;
+  
   auto &SM = Engine->SourceMgr;
   auto charRange = toCharSourceRange(SM, R);
-
+  
   // If we're replacing with something that wants spaces around it, do a bit of
   // extra work so that we don't suggest extra spaces.
```

Comment:
This probably applies to the structured fix-its too, but you can leave a FIXME for that if you want.

Reply:
I'd like to port over some more of the existing fix-its before making a decision on this. Added a FIXME for now.